### PR TITLE
Add schema validation on lisk-codec - Closes #5356

### LINKED ISSUE #5356: Add schema validation on lisk-codec
### Description

Add schema validation step before compiling the schema in lisk-codec.

### Motivation

Avoid compiling invalid schema, and expect the schema to have lisk JSON schema while encode and decoding.

### Acceptance Criteria

- Add unit test to check the invalid schema does not pass

### Additional Information
- This is not to validate the value given for encode/decode. Only for the given schema before compile

### PR DESCRIPTION
**Pull request by nazarhussain**
### What was the problem?

This PR resolves #5356

### How was it solved?

Following validation will be performed on schema: 

1. Root level schema must be of `type: 'object'`
3. Schema should have a `$id` property
4. Schema should have a `$schema` property, which should be constant value for Lisk-schema. 
4. It must have a required property available
4. All properties under schema must follow the camel case pattern 
5. For any given property either `dataType` or `type` can present. 
6. Keyword `dataType` must be among given list described in the LIP. 
7. Values for keyword `fieldNumber` should be consecutive integers for any given level of schema object. 
8. Values of keyword `filedNumber` must start from `1`. 

### How was it tested?

Run the spec tests for `lisk-codec`

### CODE REVIEW

File: elements/lisk-codec/src/schema/keywords/data_type.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import * as Debug from 'debug';
+
+// eslint-disable-next-line new-cap
+const debug = Debug('codec:keyword:dataType');
+
+export const metaSchema = {
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply elements/lisk-codec/src/schema/keywords/data_type.ts:22]: We can and we can't. I think `fieldNumber` and `dataType` is particularly related to `lisk-codec` behaviour only. So we should keep it separate.

File: elements/lisk-codec/src/schema/lisk_meta_schema.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+export const liskMetaSchema = {
+	$schema: 'http://json-schema.org/draft-07/schema#',
+	$id: 'http://lisk.io/lisk-schema/schema#',
+	title: 'Lisk Schema',
+	type: 'object',
+	properties: {
+		$schema: {
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply elements/lisk-codec/src/schema/lisk_meta_schema.ts:21]: To match `$id` and `$scehma` to be used for `validateSchema`. User don't need to provide it always, if we don't find it we override it to default `$id` of `lisk-schema`

File: elements/lisk-codec/src/schema/keywords/data_type.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import * as Debug from 'debug';
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply elements/lisk-codec/src/schema/keywords/data_type.ts:15]: Need to different in function generator and actual `debug`

**Comment by nazarhussain** [comment]
@shuse2 @pablitovicente I added bunch of unit tests for schema validation, to be clear what it is validating. Please review it particularly.

File: elements/lisk-codec/src/schema/keywords/data_type.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (shuse2) [reply elements/lisk-codec/src/schema/keywords/data_type.ts:22]: We use the validator to validate the schema, and with #5263, we need to introduce the field number and data type in the validator too

File: elements/lisk-codec/src/codec.ts [hunk 4]
```diff
@@ -22,20 +24,31 @@ import {
...
+	public addSchema(schema: Schema): ReadonlyArray<ErrorObject> {
+		const errors = validateSchema(schema);
+		if (errors.length) {
+			return errors;
+		}
+
 		const schemaName = schema.$id;
 		this._compileSchemas[schemaName] = this._compileSchema(schema, [], []);
+
+		return [];
```
Comment (shuse2) [inline elements/lisk-codec/src/codec.ts:42]: Shouldn't we just throw error here? in encode and decode it is throwing an error

Comment (nazarhussain) [reply elements/lisk-codec/src/codec.ts:42]: We can do that as well. I tried to keep that interface compliant with other validation codes where we return errors.

**Pull request merged**